lib/execution.go

```diff
@@ -0,0 +1,562 @@
+// by the provided amount.
+//
+// IMPORTANT: for UI/information purposes only, don't use for synchronization.
+func (es *ExecutionState) AddFullIterations(count uint64) uint64 {
+	return atomic.AddUint64(es.fullIterationsCount, count)
+}
+
+// GetPartialIterationCount returns the total of partial (i.e interrupted)
+// iterations that have been completed so far.
+//
+// IMPORTANT: for UI/information purposes only, don't use for synchronization.
+func (es *ExecutionState) GetPartialIterationCount() uint64 {
+	return atomic.LoadUint64(es.interruptedIterationsCount)
+}
+
+// AddInterruptedIterations increments the number of partial (i.e interrupted)
+// iterations by the provided amount.
+//
+// IMPORTANT: for UI/information purposes only, don't use for synchronization.
+func (es *ExecutionState) AddInterruptedIterations(count uint64) uint64 {
+	return atomic.AddUint64(es.interruptedIterationsCount, count)
+}
+
+// MarkStarted saves the current timestamp as the test start time.
+//
+// CAUTION: Calling MarkStarted() a second time for the same execution state will
+// result in a panic!
+func (es *ExecutionState) MarkStarted() {
+	if !atomic.CompareAndSwapInt64(es.startTime, 0, time.Now().UnixNano()) {
+		panic("the execution scheduler was started a second time")
+	}
+}
+
+// MarkEnded saves the current timestamp as the test end time.
+//
+// CAUTION: Calling MarkEnded() a second time for the same execution state will
+// result in a panic!
+func (es *ExecutionState) MarkEnded() {
+	if !atomic.CompareAndSwapInt64(es.endTime, 0, time.Now().UnixNano()) {
+		panic("the execution scheduler was stopped a second time")
+	}
+}
+
+// HasStarted returns true if the test has actually started executing.
+// It will return false while a test is in the init phase, or if it has
+// been initially paused. But if will return true if a test is paused
+// midway through its execution (see above for details regarind the
+// feasibility of that pausing for normal executors).
+func (es *ExecutionState) HasStarted() bool {
+	return atomic.LoadInt64(es.startTime) != 0
+}
+
+// HasEnded returns true if the test has finished executing. It will return
+// false until MarkEnded() is called.
+func (es *ExecutionState) HasEnded() bool {
+	return atomic.LoadInt64(es.endTime) != 0
+}
+
+// IsPaused quickly returns whether the test is currently paused, by reading
+// the atomic currentPauseTime timestamp
+func (es *ExecutionState) IsPaused() bool {
+	return atomic.LoadInt64(es.currentPauseTime) != 0
+}
+
+// GetCurrentTestRunDuration returns the duration for which the test has already
+// ran. If the test hasn't started yet, that's 0. If it has started, but has
+// been paused midway through, it will return the time up until the pause time.
+// And if it's currently running, it will return the time since the start time.
+//
+// IMPORTANT: for UI/information purposes only, don't use for synchronization.
+func (es *ExecutionState) GetCurrentTestRunDuration() time.Duration {
```

Comment:
Nit: could this be shortened to `GetTestRunDuration()`? When is it not `Current`?

Reply:
I think the idea is that it is 'Current' in the sense that how much it has ran so far, not how much it will run in total. There isn't any `GetTestRunDuration` but I would expect just from the name for it to return how long it would run, not how long it had ran.

Reply:
@MStoykov OK, that makes sense, thanks.